platforms/defaults_windows.go

```diff
@@ -92,3 +99,27 @@ func prefix(v string) string {
 func Default() MatchComparer {
 	return Only(DefaultSpec())
 }
+
+var (
+	ubr  int
+	once sync.Once
+)
+
+func GetCurrentWindowsUpdateBuildRevision() int {
```

Comment:
Can we not use `windows.GetVersion` how is this different?

Reply:
[`windows.GetVersion`](https://pkg.go.dev/golang.org/x/sys/windows#GetVersion) (which calls [`kernel32.GetVersion`](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getversion)) lies about the current version, if you don't pull in a manifest. I don't believe we pull [the manifest](https://github.com/microsoft/hcsshim/blob/4b67ed2cdb5355b30cf572ec17d5ffe648a85b0e/osversion/osversion_windows.go) into the containerd daemon builds, just the test binaries.

Reply:
The right fix isn't to start reading from the registry lol, its to manifest the binary. Am I wrong?

Reply:
Well, as I understand (but have not myself confirmed), e.g. `cmd /c ver` is reading the registry to get the value, despite being bundled with the OS.

I also note from the documentation of `kernel32.GetVersion` that it won't return versions newer than the version you manifest for. Windows 10 and 11 are [counted as the same version in this context](https://docs.microsoft.com/en-us/windows/win32/sysinfo/targeting-your-application-at-windows-8-1), but I don't know if it's reliable that for the future.

I also don't know what the "report Windows 10 when actually running Windows 15" logic does to the build number. It's quite possible that today it uses the current UBR and will hence report a non-existent version that would be perfect for our use-case here (since we get the rest of the version number from a different API).

Basically, my impression is that the platform owners are trying to kill that API, and I don't trust them not to make it do "documented but whacky" things down the track, even with a manifest that makes it reasonably non-whacky today.

AFAIK, unlike the rest of the version number, the UBR is really just a tracking point for Windows Update: the registry _is_ the canonical record because any given update doesn't guarantee updating any particular file's version, e.g. my current UBR is 1586, but my kernel32.dll is 1566.

If someone with access to the Windows source code looked and told me `kernel32.GetVersion` read the UBR from the registry, I'd be 0% surprised.

Reply:
We'd had some talks about moving off of `GetVersion` as ensuring anything that calls the github.com/microsoft/hcsshim/osversion package is not fun (and from talking with some OS folks the registry is generally regarded as a stable source for this info). I see a bunch of spots it's used in the OS, but I can't confidently speak on if this is the "recommended" approach. Let me ask around some more tomorrow to see.

Reply:
Windows Terminal for a quick example manifests their binary (and makes use of [VerifyVersionInfoW](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-verifyversioninfow in a couple places to check functionality availability https://github.com/microsoft/terminal/blob/70d8b2a2ef70cd4148759e695d6a9eeee8c94764/src/cascadia/TerminalSettingsModel/WslDistroGenerator.cpp#L36). 

From talking with a colleague on the console team he recommends the manifesting approach, but for the UBR I don't know how you'd get around not getting it from the registry..

Reply:
Since this is on the table, I thought I'd mention https://github.com/moby/moby/pull/43431 in passing, which is implementing a nice, go-native manifest generator using https://github.com/josephspurrier/goversioninfo. And so perhaps this could be used to make `github.com/microsoft/hcsshim/osversion` do the right thing in general using `go:generate`?

I haven't look too hard at this yet, but I know that having to pull on `github.com/microsoft/hcsshim/tests/functional/manifest` is awkward, and now someone has done the "someday" work of implementing a windres for `go:generate`.